lib/spack/llnl/util/filesystem.py

```diff
@@ -924,6 +924,71 @@ def traverse_tree(source_root, dest_root, rel_path='', **kwargs):
         yield (source_path, dest_path)
 
 
+def visit_directory_tree(root, visitor, rel_path='', depth=0):
+    """
+    Recurses the directory root depth-first through a visitor pattern
+
+    The visitor interface is as follows:
+    - handle_file(root, rel_path, depth) -> bool
+        if False is returned, the iteration over children in the current dir is stopped
+    - before_enter_dir(root, rel_path, depth) -> bool
+        if True is returned, descends into this directory
+    - before_enter_symlinked_dir(root, rel_path, depth) -> bool
+        if True is returne,d descends into this directory
+    - after_enter_dir(root, rel_path, depth) -> void
+        only called when before_enter_dir returns True
+    - after_enter_symlinked_dir(root, rel_path, depth) -> void
+        only called when before_enter_symlinked_dir returns True
+    """
+    dir = os.path.join(root, rel_path)
+
+    for f in os.listdir(dir):
+        rel_child = os.path.join(rel_path, f)
+        child = os.path.join(dir, f)
+
+        st, lst = None, None
+        is_link, is_dir = False, False
+
+        # lstat first to avoid having to stat when not a symlink
+        try:
+            lst = os.lstat(child)
+            is_link = stat.S_ISLNK(lst.st_mode)
+        except OSError:
+            pass
+
+        # if lstat succeeded and we don't have a link, check wether we have a dir
+        if lst is not None and not is_link:
+            is_dir = stat.S_ISDIR(lst.st_mode)
+
+        # if lstat succeeded and we have a symlink, we must verify it's a dir
+        # through a stat call.
+        if lst is not None and is_link:
+            # if it was a symlink, do an actual stat
+            try:
+                st = os.stat(child)
+                is_dir = stat.S_ISDIR(st.st_mode)
+            except OSError:
+                pass
```

Comment:
```suggestion
        # if lstat succeeded and we have a symlink, we must verify it's a dir
        # through a stat call.
        if is_link:
            # if it was a symlink, do an actual stat
            try:
                st = os.stat(child)
                is_dir = stat.S_ISDIR(st.st_mode)
            except OSError:
                pass
        else:
            is_dir = stat.S_ISDIR(lst.st_mode)
```
Since we can continue if lst failed